realm/src/main/java/io/realm/BaseRealm.java

```diff
@@ -586,14 +626,278 @@ public static synchronized void migrateRealm(RealmConfiguration configuration, R
     private class RealmCallback implements Handler.Callback {
         @Override
         public boolean handleMessage(Message message) {
-            if (message.what == REALM_CHANGED) {
-                sharedGroupManager.advanceRead();
-                sendNotifications();
+            switch (message.what) {
+                case REALM_CHANGED: {
+                    if (threadContainsAsyncQueries()) {
+                        if (updateAsyncQueriesTask != null && !updateAsyncQueriesTask.isDone()) {
+                            // try to cancel any pending update since we're submitting a new one anyway
+                            updateAsyncQueriesTask.cancel(true);
+                            asyncQueryExecutor.getQueue().remove(updateAsyncQueriesTask);
+                            RealmLog.d("REALM_CHANGED realm:"+ BaseRealm.this + " cancelling pending REALM_UPDATE_ASYNC_QUERIES updates");
+                        }
+                        RealmLog.d("REALM_CHANGED realm:"+ BaseRealm.this + " updating async queries, total: " + asyncRealmResults.size());
+                        // prepare a QueryUpdateTask to current async queries in this thread
+                        QueryUpdateTask.Builder.UpdateQueryStep updateQueryStep = QueryUpdateTask.newBuilder()
+                                .realmConfiguration(getConfiguration());
+                        QueryUpdateTask.Builder.RealmResultsQueryStep realmResultsQueryStep = null;
+
+                        // we iterate over non GC'd async RealmResults then add them to the list to be updated (in a batch)
+                        Iterator<Map.Entry<WeakReference<RealmResults<?>>, RealmQuery<?>>> iterator = asyncRealmResults.entrySet().iterator();
+                        while (iterator.hasNext()) {
+                            Map.Entry<WeakReference<RealmResults<?>>, RealmQuery<?>> entry = iterator.next();
+                            WeakReference<RealmResults<?>> weakReference = entry.getKey();
+                            RealmResults<?> realmResults = weakReference.get();
+                            if (realmResults == null) {
+                                // GC'd instance remove from the list
+                                iterator.remove();
+
+                            } else {
+                                realmResultsQueryStep = updateQueryStep.add(weakReference,
+                                        entry.getValue().handoverQueryPointer(),
+                                        entry.getValue().getArgument());
+                            }
+
+                            // Note: we're passing an WeakRef of a RealmResults to another thread
+                            //       this is safe as long as we don't invoke any of the RealmResults methods.
```

Comment:
if you use the hashCode you need another HashMap to do book keeping between hashCode & instances. IMO this is fine we internally the thread model of our RealmResults